connectome_interpreter/utils.py

```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+
+    Args:
+        grid_outputs (torch.Tensor|np.ndarray): The output grid tensor of shape (grid_size**2, num_neurons, num_layers).
+        grid_coords (list): A list of grid coordinates. Best from function `make_grid_inputs()`.
+        selected_index (int|list): The index or indices to plot.
+        figsize (tuple, optional): The size of the figure. Defaults to (10, 8).
+    """
+    selected_index = to_nparray(selected_index)
+
+    # if grid_outputs is tensor, convert to numpy
+    if isinstance(grid_outputs, torch.Tensor):
+        grid_outputs = grid_outputs.cpu().numpy()
+
+    if title is None:
+        title = f'Output Grid for {selected_index}'
+
+    def plot_heatmap(index, xlab, ylab, title):
+        # # Create the plot
+        plt.figure(figsize=figsize)
+        im = plt.imshow(heatmaps[index-1], cmap=cmap, origin='lower',
+                        extent=[0, 1, 0, 1])
+        plt.colorbar(im, label='Output')
+        plt.xlabel(xlab)
+        plt.ylabel(ylab)
+        plt.title(title)
+        plt.show()
+
+    heatmaps = []
+
+    for i in range(grid_outputs.shape[2]):
+        # Extract the selected index values
+        values = grid_outputs[:, selected_index,
+                              i].mean(axis=1)
+
+        # Reshape values to match the grid
+        grid_size = int(np.sqrt(len(grid_coords)))
+        values_grid = values.reshape((grid_size, grid_size))
+        heatmaps.append(values_grid)
+
+    slider = widgets.IntSlider(value=1, min=1, max=len(heatmaps), step=1,
+                               description='Timepoint', continuous_update=True)
+
+    # Link the slider to the plotting function
+    display(widgets.interactive(plot_heatmap, index=slider,
+            xlab=xlab, ylab=ylab, title=title))
+
+
+@dataclass
+class XORCircuit:
+    # Input nodes
+    input1: list
+    input2: list
+    # Middle layer nodes
+    exciter1: str  # Takes input from input1 only
+    exciter2: str  # Takes input from input2 only
+    inhibitor: str        # Takes input from both inputs
+    # Output node
+    output: list
+
+
+def find_xor(paths: pd.DataFrame) -> List[XORCircuit]:
+    """
+    Find XOR-like circuits in a 3-layer network, based on [Wang et al. 2024](https://www.biorxiv.org/content/10.1101/2024.09.24.614724v2). Note: this function currently ignores middle excitatory neruons that receive both inputs.
+
+    Args:
+        paths: DataFrame with columns ['pre', 'post', 'sign', 'layer']
+        pre: source node
+        post: target node
+        sign: 1 (excitatory) or -1 (inhibitory)
+        layer: 1 (input->middle) or 2 (middle->output)
+
+    Returns:
+    List of XORCircuit objects, each representing a found XOR motif
+    """
+    # checking input ----
+    if set(paths.layer) != {1, 2}:
+        raise ValueError(
+            "The input DataFrame should have exactly 2 unique layers, 1 and 2")
+
+    # check column names of paths
+    if not {'pre', 'post', 'sign', 'layer'}.issubset(paths.columns):
```

Comment:
Optionally `not s1 <= s2` instead of `not s1.issubset(s2)`. Although it can be confusing since `not s1 <= s2` isn't the same thing as `s2 < s1`!